src/key_gen/tests.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright 2020 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::dev_utils::{create_ids, PeerId};
+use crate::id::SecretId;
+use crate::key_gen::{message::DkgMessage, DkgPhases, KeyGen};
+use bincode::serialize;
+use rand::Rng;
+use std::collections::{BTreeMap, BTreeSet};
+
+// Alter the configure of the number of nodes and the threshold.
+const NODENUM: usize = 7;
+const THRESHOLD: usize = 5;
+
+fn setup_generators(
+    mut rng: &mut dyn Rng,
+    dkg_phase: DkgPhases,
+    non_responsive: Option<usize>,
+) -> (Vec<PeerId>, Vec<KeyGen<PeerId>>) {
+    // Generate individual ids and key pairs.
+    let peer_ids: Vec<PeerId> = create_ids(NODENUM);
+    let pub_keys: BTreeSet<PeerId> = peer_ids.iter().cloned().collect();
+
+    // Create the `KeyGen` instances
+    let mut generators = Vec::new();
+    let mut proposals = Vec::new();
+    peer_ids.iter().enumerate().for_each(|(index, peer_id)| {
+        let mut key_gen = if dkg_phase == DkgPhases::Initialization {
+            let (key_gen, proposal) =
+                KeyGen::<PeerId>::initialize(&peer_id.sec_key(), THRESHOLD, pub_keys.clone())
+                    .unwrap_or_else(|_err| panic!("Failed to initial KeyGen of {:?}", &peer_id));
+            proposals.push(proposal);
+            key_gen
+        } else {
+            KeyGen::<PeerId>::initialize_for_test(
+                peer_id.public_id().clone(),
+                index as u64,
+                pub_keys.clone(),
+                THRESHOLD,
+                dkg_phase,
+            )
+        };
+
+        // Calling `finalize_complaining_phase` to trigger the key_gen instance transit into
+        // Justification phase and send out the initial proposal.
+        if dkg_phase == DkgPhases::Complaining {
+            let proposal = key_gen
+                .finalize_complaining_phase(&peer_id.sec_key(), &mut rng)
+                .unwrap_or_else(|_err| {
+                    panic!("Failed to complete complaint stage of {:?}", &peer_id)
+                });
+            proposals.push(proposal);
+        }
+        generators.push(key_gen);
+    });
+
+    messaging(
+        &mut rng,
+        &peer_ids,
+        &mut generators,
+        &mut proposals,
+        non_responsive,
+    );
+
+    (peer_ids, generators)
+}
+
+fn messaging(
+    mut rng: &mut dyn Rng,
+    peer_ids: &[PeerId],
+    generators: &mut Vec<KeyGen<PeerId>>,
+    proposals: &mut Vec<DkgMessage<PeerId>>,
```

Comment:
Why is this called `proposals` instead of `messages` (also elsewhere)?

Reply:
just name it as proposal based on the `purpose of the message`
(naming of `messages` sounds too plain :)  )
I don't mind change to `messages` if you insisit.   :)

Reply:
OK, makes sense. No need to rename it.